invoker/core/src/main/java/com/google/cloud/functions/invoker/TypedFunctionExecutor.java

```diff
@@ -0,0 +1,195 @@
+import javax.servlet.http.HttpServletResponse;
+
+public class TypedFunctionExecutor extends HttpServlet {
+  private static final String APPLY_METHOD = "apply";
+  private static final Logger logger = Logger.getLogger("com.google.cloud.functions.invoker");
+
+  private final Type argType;
+  private final TypedFunction<Object, Object> func;
+  private final ConfigurationImpl configuration;
+
+  private TypedFunctionExecutor(
+      Type argType, TypedFunction<Object, Object> func, ConfigurationImpl configuration) {
+    this.argType = argType;
+    this.func = func;
+    this.configuration = configuration;
+  }
+
+  public static TypedFunctionExecutor forClass(Class<?> genericClass) {
+    if (!TypedFunction.class.isAssignableFrom(genericClass)) {
+      throw new RuntimeException(
+          "Class "
+              + genericClass.getName()
+              + " does not implement "
+              + TypedFunction.class.getName());
+    }
+    @SuppressWarnings("unchecked")
+    Class<? extends TypedFunction<?, ?>> clazz =
+        (Class<? extends TypedFunction<?, ?>>) genericClass.asSubclass(TypedFunction.class);
+
+    Optional<Type> argType = TypedFunctionExecutor.handlerTypeArgument(clazz);
+    if (argType.isEmpty()) {
+      throw new RuntimeException(
+          "Class " + clazz.getName() + " does not implement " + TypedFunction.class.getName());
+    }
+
+    TypedFunction<?, ?> funcInstance;
+    try {
+      funcInstance = clazz.getDeclaredConstructor().newInstance();
+    } catch (Exception e) {
+      throw new RuntimeException(
+          "Class "
+              + clazz.getName()
+              + " must declare a valid default constructor to be usable as a strongly typed"
+              + " function. Could not use constructor: "
+              + e.toString());
+    }
+
+    ConfigurationImpl configuration = new ConfigurationImpl();
+    funcInstance.configure(configuration);
+
+    @SuppressWarnings("unchecked")
+    TypedFunctionExecutor executor =
+        new TypedFunctionExecutor(
+            argType.orElseThrow(), (TypedFunction<Object, Object>) funcInstance, configuration);
+    return executor;
+  }
+
+  /**
+   * Returns the {@code T} of a concrete class that implements {@link BackgroundFunction
+   * BackgroundFunction<T>}. Returns an empty {@link Optional} if {@code T} can't be determined.
+   */
+  static Optional<Type> handlerTypeArgument(Class<? extends TypedFunction<?, ?>> functionClass) {
+    return Arrays.stream(functionClass.getMethods())
+        .filter(method -> method.getName().equals(APPLY_METHOD) && method.getParameterCount() == 1)
+        .map(method -> method.getGenericParameterTypes()[0])
+        .findFirst();
+  }
+
+  /** Executes the user's method, can handle all HTTP type methods. */
+  @Override
+  public void service(HttpServletRequest req, HttpServletResponse res) {
+    HttpRequestImpl reqImpl = new HttpRequestImpl(req);
+    HttpResponseImpl respImpl = new HttpResponseImpl(res);
```

Comment:
nit - the other code does call this respImpl, though I think that's internally inconsistent since it chose to use 'res'... I would correct that here and use resImpl. Ideally we would just be using proper java style naming like "request" and "requestImpl"...